sd_webui_bayesian_merger/merger.py

```diff
@@ -124,83 +144,62 @@ def merge_key(
+                current_bases = {k: w[weight_index] for k, w in weights.items()}
+
+            merged = self.merge_block(current_bases, thetas, key)
 
-        merged = self.merge_block(
-            c_alpha,
-            c_beta,
-            theta_0[key],
-            theta_1[key],
-            theta_2[key] if self.cfg.merge_mode != "weighted_sum" else None,
-            key in theta_2 if self.cfg.merge_mode != "weighted_sum" else None,
-        )
+            if not best or self.cfg.best_precision == "16":
+                merged = merged.half()
 
-        if not best or self.cfg.best_precision == "16":
-            merged = merged.half()
+            return (key, merged)
 
-        return (key, merged)
+    def merge_block(self, current_bases: Dict, thetas: Dict, key: str) -> Dict:
+        t0 = thetas["model_a"][key]
+        t1 = thetas["model_b"][key]
+        alpha = current_bases["alpha"]
+        if self.cfg.merge_mode == "weighted_sum":
+            return (1 - alpha) * t0 + alpha * t1
 
-    def merge_block(
-        self,
-        alpha: float,
-        beta: float,
-        t0: Dict,
-        t1: Dict,
-        t2: Dict,
-        k_in_t2: bool,
-    ) -> Dict:
+        t2 = thetas["model_c"][key]
         if self.cfg.merge_mode == "add_difference":
-            return t0 + alpha * (t1 - t2) if k_in_t2 else t0
-        elif self.cfg.merge_mode == "sum_twice":
-            merged = (1 - alpha) * t0 + alpha * t1
-            return (1 - beta) * merged + beta * t2
+            return t0 + alpha * (t1 - t2)
+        elif self.cfg.merge_mode == "weighted_add_difference":
+            return (1 - alpha) * t0 + alpha * (t1 - t2)
+
+        beta = current_bases["beta"]
+        if self.cfg.merge_mode == "sum_twice":
+            return (1 - beta) * ((1 - alpha) * t0 + alpha * t1) + beta * t2
         elif self.cfg.merge_mode == "triple_sum":
             return (1 - alpha - beta) * t0 + alpha * t1 + beta * t2
-        elif self.cfg.merge_mode == "weighted_sum":
-            return (1 - alpha) * t0 + alpha * t1
+        elif self.cfg.merge_mode == "weighted_double_difference":
+            t3 = thetas["model_d"][key]
+            t4 = thetas["model_e"][key]
+            return (1 - alpha) * t0 + alpha * (
+                (1 - beta) * (t1 - t2) + beta * (t3 - t4)
+            )
 
     def merge(
         self,
-        weights_alpha: List[float],
-        weights_beta: List[float],
-        base_alpha: float,
-        base_beta: float,
+        weights: Dict,
+        bases: Dict,
         best: bool = False,
     ) -> None:
-        if len(weights_alpha) != NUM_TOTAL_BLOCKS:
-            raise ValueError(f"weights value must be {NUM_TOTAL_BLOCKS}")
+        thetas = {k: self.load_sd_model(m) for k, m in self.models.items()}
+        all_keys = [list(set(t.keys())) for t in thetas.values()]
+        pair_of_keys = all_keys.pop()
+        assert all([k == pair_of_keys for k in all_keys])
```

Comment:
This seems much too strong.

Reply:
Actually, I think the base assertion/comparison is wrong, because even when any is used instead of all, it still throws an asserterror.

Reply:
It's tough for me to test at the moment as it takes 39200 iterations just to complete the first part of the merge process. It's basically 16-17 hours to complete a merge, (at 10 init points, 10 n_iter, 10 batch size, and 13 payloads). I'll probably lower it down to 5 batch size for the next run I do.

Reply:
Yeah, once the previous merge finishes, I'll test the new version with a smaller batch size.